File: python/flyql/generators/starrocks/column.py
"""StarRocks-dialect Column."""

import re
from typing import Dict, List, Optional

from flyql.flyql_type import Type

REGEX = {
    Type.String: re.compile(r"^(varchar|char|string)\s*\(\s*\d+\s*\)"),
    Type.Int: re.compile(r"^(tinyint|smallint|int|largeint|bigint)\s*\(\s*\d+\s*\)"),
    Type.Float: re.compile(r"^(decimal|float|double)\d*\s*\(\s*\d+\s*(,\s*\d+)?\s*\)"),
    Type.Date: re.compile(r"^datetime"),
    Type.Array: re.compile(r"^array\s*\<"),
    Type.Map: re.compile(r"^map\s*\<"),
    Type.Struct: re.compile(r"^struct\s*\<"),
    Type.JSON: re.compile(r"^json"),
}

# SR renames: special→unknown.
FLYQL_TYPE_TO_STARROCKS_TYPES: Dict[Type, set] = {
    Type.String: {"string", "varchar", "char", "binary", "varbinary"},
    Type.Int: {"int", "tinyint", "smallint", "largeint", "bigint"},
    Type.Float: {"float", "double", "decimal"},
    Type.Bool: {"bool", "boolean"},
    Type.Date: {"date", "datetime"},
    Type.Unknown: {"bitmap", "hll"},
    Type.JSON: {"json"},
}


def normalize_starrocks_type(sr_type: str) -> Type:
    if not sr_type or not isinstance(sr_type, str):
        return Type.Unknown

    normalized = sr_type.strip().lower()

    if normalized == "jsonstring":
        return Type.JSONString

    if REGEX[Type.String].match(normalized):
        return Type.String
    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.String]:
        return Type.String

    if REGEX[Type.Int].match(normalized):
        return Type.Int
    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.Int]:
        return Type.Int

    if REGEX[Type.Float].match(normalized):
        return Type.Float
    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.Float]:
        return Type.Float

    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.Bool]:
        return Type.Bool

    if REGEX[Type.Date].match(normalized):
        return Type.Date
    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.Date]:
        return Type.Date

    if REGEX[Type.JSON].match(normalized):
        return Type.JSON
    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.JSON]:
        return Type.JSON

    if REGEX[Type.Array].match(normalized):
        return Type.Array

    if REGEX[Type.Map].match(normalized):
        return Type.Map

    if REGEX[Type.Struct].match(normalized):
        return Type.Struct

    if normalized in FLYQL_TYPE_TO_STARROCKS_TYPES[Type.Unknown]:
        return Type.Unknown

    return Type.Unknown


class Column:
    """Opaque StarRocks-dialect column."""

    def __init__(
        self,
        name: str,
        _type: str,
        values: Optional[List[str]] = None,
        display_name: str = "",
        raw_identifier: str = "",
    ):
        self.name = name
        self.match_name = name
        self.values: List[str] = values or []
        self.display_name = display_name
        self.raw_identifier = raw_identifier
        self._raw_type = _type
        self._flyql_type: Type = normalize_starrocks_type(_type)

    @property
    def raw_type(self) -> str:
        return self._raw_type

    @property
    def flyql_type(self) -> Type:
        return self._flyql_type

    def with_raw_identifier(self, identifier: str) -> "Column":
        self.raw_identifier = identifier
        return self

File: javascript/packages/flyql/src/generators/starrocks/column.js
import { Type } from '../../flyql_type.js'
import { FlyqlError } from '../../core/exceptions.js'

const typeRegexes = {
    [Type.String]: /^(varchar|char|string)\s*\(\s*\d+\s*\)/i,
    [Type.Int]: /^(tinyint|smallint|int|largeint|bigint)\s*\(\s*\d+\s*\)/i,
    [Type.Float]: /^(decimal|float|double)\d*\s*\(\s*\d+\s*(,\s*\d+)?\s*\)/i,
    [Type.Date]: /^datetime/i,
    [Type.Array]: /^array\s*</i,
    [Type.Map]: /^map\s*</i,
    [Type.Struct]: /^struct\s*</i,
    [Type.JSON]: /^json/i,
}

const flyqlTypeToStarRocksTypes = {
    [Type.String]: new Set(['string', 'varchar', 'char', 'binary', 'varbinary']),
    [Type.Int]: new Set(['int', 'tinyint', 'smallint', 'largeint', 'bigint']),
    [Type.Float]: new Set(['float', 'double', 'decimal']),
    [Type.Bool]: new Set(['bool', 'boolean']),
    [Type.Date]: new Set(['date', 'datetime']),
    [Type.Unknown]: new Set(['bitmap', 'hll']),
    [Type.JSON]: new Set(['json']),
}

export function normalizeStarRocksType(srType) {
    if (!srType) return Type.Unknown
    const normalized = srType.trim().toLowerCase()

    if (normalized === 'jsonstring') return Type.JSONString

    if (typeRegexes[Type.String].test(normalized)) return Type.String
    if (flyqlTypeToStarRocksTypes[Type.String].has(normalized)) return Type.String

    if (typeRegexes[Type.Int].test(normalized)) return Type.Int
    if (flyqlTypeToStarRocksTypes[Type.Int].has(normalized)) return Type.Int

    if (typeRegexes[Type.Float].test(normalized)) return Type.Float
    if (flyqlTypeToStarRocksTypes[Type.Float].has(normalized)) return Type.Float

    if (flyqlTypeToStarRocksTypes[Type.Bool].has(normalized)) return Type.Bool

    if (typeRegexes[Type.Date].test(normalized)) return Type.Date
    if (flyqlTypeToStarRocksTypes[Type.Date].has(normalized)) return Type.Date

    if (typeRegexes[Type.JSON].test(normalized)) return Type.JSON
    if (flyqlTypeToStarRocksTypes[Type.JSON].has(normalized)) return Type.JSON

    if (typeRegexes[Type.Array].test(normalized)) return Type.Array
    if (typeRegexes[Type.Map].test(normalized)) return Type.Map
    if (typeRegexes[Type.Struct].test(normalized)) return Type.Struct

    if (flyqlTypeToStarRocksTypes[Type.Unknown].has(normalized)) return Type.Unknown

    return Type.Unknown
}

/**
 * Opaque StarRocks-dialect Column. Construct via `newColumn({...})` or `new Column({...})`.
 * The flyql semantic type is computed at construction via `normalizeStarRocksType`.
 */
export class Column {
    constructor(opts) {
        if (opts == null || typeof opts !== 'object' || Array.isArray(opts)) {
            throw new FlyqlError(`Column: expected an options object ({ name, type, values? }); got ${typeof opts}`)
        }
        const { name, type, values = null, displayName = '', rawIdentifier = '' } = opts
        if (typeof name !== 'string' || !name) {
            throw new FlyqlError(`Column: 'name' must be a non-empty string`)
        }
        if (typeof type !== 'string') {
            throw new FlyqlError(
                `Column: 'type' must be a raw-type string (e.g. 'VARCHAR', 'jsonstring'); got ${typeof type}`,
            )
        }

        this.name = name
        this.values = values || []
        this.displayName = displayName
        this.rawIdentifier = rawIdentifier
        this.matchName = name
        this._rawType = type
        this._flyqlType = normalizeStarRocksType(type)
    }

    rawType() {
        return this._rawType
    }

    flyqlType() {
        return this._flyqlType
    }

    /** @deprecated Pass `rawIdentifier` in the options object to `newColumn()` / `new Column()` instead. Slated for removal in a follow-up. */
    withRawIdentifier(identifier) {
        this.rawIdentifier = identifier
        return this
    }
}

export function newColumn(opts) {
    if (opts == null || typeof opts !== 'object' || Array.isArray(opts)) {
        throw new FlyqlError(`newColumn: expected an options object ({ name, type, values? }); got ${typeof opts}`)
    }
    return new Column(opts)
}
